pkg/trtl/trtl_test.go

```diff
@@ -1,39 +1,323 @@
+		}
+	}
+}
+
+// writeGzip writes the database in the temporary directory to a gzipped file.
+func writeGzip(s *trtlTestSuite) {
+	require := s.Require()
+	// Create a gzip file.
+	f, err := os.Create(s.gzip)
+	require.NoError(err)
+	defer f.Close()
+	w := gzip.NewWriter(f)
+	defer w.Close()
+
+	// Create a tar file.
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	// Write the DB to the tar file.
+	err = filepath.Walk(s.db, func(path string, info os.FileInfo, err error) error {
+		require.NoError(err)
+		hdr, err := tar.FileInfoHeader(info, "")
+		require.NoError(err)
+		hdr.Name = path[len(s.db):]
+		err = tw.WriteHeader(hdr)
+		require.NoError(err)
+		if info.IsDir() {
+			return nil
+		}
+		f, err := os.Open(path)
+		require.NoError(err)
+		defer f.Close()
+		_, err = io.Copy(tw, f)
+		require.NoError(err)
+		return nil
+	})
+	require.NoError(err)
+}
+
+func (s *trtlTestSuite) SetupSuite() {
+	var err error
+	require := s.Require()
+	s.gzip = filepath.Join("testdata", "db.tar.gz")
+	s.db, err = ioutil.TempDir("testdata", "db*")
+	require.NoError(err)
+
+	// Only regenerate the test database if requested or it doesn't exist.
+	if _, err = os.Stat(s.gzip); *update || os.IsNotExist(err) {
+		generateDB(s)
+	} else {
+		loadFixtures(s)
 	}
 
-	server, err := trtl.New(config)
-	require.NoError(t, err)
+	// Always extract the test database to a temporary directory.
+	extractGzip(s)
+
+	// Load default config and add database path.
+	os.Setenv("TRTL_DATABASE_URL", "leveldb:///"+s.db)
+	os.Setenv("TRTL_PID", "1")
+	os.Setenv("TRTL_REGION", "foo")
+	s.conf, err = config.New()
+	require.NoError(err)
+}
+
+func (s *trtlTestSuite) TearDownSuite() {
+	require := s.Require()
+	err := os.RemoveAll(s.db)
+	require.NoError(err)
+}
+
+func TestTrtl(t *testing.T) {
+	suite.Run(t, new(trtlTestSuite))
+}
+
+// Test that we can call the Get RPC and get the correct response.
+func (s *trtlTestSuite) TestGet() {
+	var actual interface{}
+
+	require := s.Require()
+	alice := dbFixtures["alice"]
+	object := dbFixtures["object"]
 
-	err = server.Serve()
-	require.NoError(t, err)
+	// Start the server.
+	server, err := trtl.New(s.conf)
+	require.NoError(err)
+	go server.Serve()
+	defer server.Shutdown()
 
-	// Test that we can get a response from a gRPC request.
+	// Start the gRPC client.
 	ctx := context.Background()
-	conn, err := grpc.DialContext(ctx, "localhost:1313", grpc.WithInsecure())
-	require.NoError(t, err)
+	conn, err := grpc.DialContext(ctx, "localhost"+s.conf.BindAddr, grpc.WithInsecure())
+	require.NoError(err)
+	defer conn.Close()
 	client := pb.NewTrtlClient(conn)
-	_, err = client.Get(ctx, &pb.GetRequest{Key: []byte("foo")})
-	require.EqualError(t, err, "rpc error: code = Unimplemented desc = not implemented")
 
-	err = server.Shutdown()
-	require.NoError(t, err)
+	// Retrieve a value from a reserved namespace - should fail.
+	_, err = client.Get(ctx, &pb.GetRequest{
+		Namespace: "default",
+		Key:       []byte(object.Key),
+	})
```

Comment:
Thank you for adding test coverage for the new reserved namespaces we introduced in the last merge!